Pkl/Decoder/Decoder.cs

```diff
@@ -2,12 +2,74 @@
 
 namespace Pkl.Decoding;
 
-public class Decoder
+public partial class Decoder
 {
-    public object Decode(byte[] input)
+    public T Decode<T>(byte[] input) where T : notnull
     {
         var reader = new MessagePackReader(input);
-        // TODO
-        return null!;
+        
+        return DecodeAny<T>(ref reader);
+    }
+
+    private T DecodeAny<T>(ref MessagePackReader reader) where T : notnull
+    {
+        var value = DecodeAny(ref reader, typeof(T));
+
+        if (value is T valueT)
+        {
+            return valueT;
+        }
+
+        throw new Exception("Invalid type parameter " + nameof(T));
+    }
```

Comment:
Have you considered the `TryX` pattern (like TryParse, TryGet) where the signature is like so:
```
private bool TryDecodeAny<T>(ref MessagePackReader reader, out T result);
```
Where the bool indicates true on a success and false on a failure. The `result` value gets set to null on a failure and to the actual result on a success.

The main reason I query this is that there are a number of examples in c# with this pattern already and would seem a bit easier to work with than handling an exception.

Example of a TryParse: https://learn.microsoft.com/en-us/dotnet/api/system.int32.tryparse?view=net-8.0

Reply:
Yeah, now that you've mentioned it, I also think having a Try[...] API would make it easier to work with. Here's what I'm thinking of:

Instead of making a TryDecodeAny method (or making the other Decoder methods into TryX), making a Try counterpart to the Evaluate[...] methods on `IEvaluator`. So, for instance:

The existing `Task<T> EvaluateModule<T>(ModuleSource source)` would have its `Task<bool> TryEvaluateModule<T>(ModuleSource source, out T result)` counterpart, which would essentially wrap a call to `EvaluateModule<T>` in a try/catch block and return false in the case of it catching an exception.

The reason why I'm tending towards keeping the non Try[...] versions of the methods is because the exceptions give context as to why the decoding didn't succeed.

Does that make sense to you? I'm not sure if that's the best way to go about it.

Reply:
Yeah that makes sense, didn't realise async was an issue